### ISSUE TITLE
AIRL Replay buffer

### ISSUE THREAD

**Issue by shwang** [HumanCompatibleAI/imitation#17]
Not implemented yet:

> RL methods commonly learn rewards which explain behavior locally for the current policy, because the reward can "forget" the signal that it gave to an earlier policy. This makes rewards obtained at the end of training difficult to optimize from scratch, as they overfit to samples from the current iteration.

> To somewhat mitigate this effect, we mix policy samples from the previous 20 iterations of training as negatives when training the discriminator

**Comment by shwang** [comment]
To elaborate, we need to add a replay buffer containing generator policy samples from the last 20 iterations/steps of training.

In terms of PPO, I'm pretty sure that this means replay from the past 20 "epochs", as [epochs are defined here](https://github.com/HumanCompatibleAI/imitation/blob/master/imitation/airl.py#L121). (Or maybe "iterations" means something more intermediate here? My knowledge of PPO and Justin's codebase is low.)

@AdamGleave @qxcv Do either of you know?

**Comment by qxcv** [comment]
Yes, epochs. Also, we discussed this yesterday, but it's worth not taking the stuff in the paper too seriously b/c the actual code tends to work differently (e.g I'm guessing that just making a replay buffer with capacity for `20 * rollouts_per_epoch * max_rollout_len` samples will suffice).

**Comment by shwang** [comment]
Is there a standard Relay buffer utility? I’m thinking of just using a
queue like thing for the last 20 epochs and concatenating the 20 inputs for
now (on the numpy side)

**Comment by qxcv** [comment]
I don't think there's one in the imitation repo yet. What you have in mind sounds good though :+1:

**Comment by shwang** [comment]
I could be slightly more efficient with a circular buffer thing but I don’t
thinking it’s worth bothering yet

**Comment by shwang** [comment]
Sounds good, thanks

**Comment by AdamGleave** [comment]
Maybe a bit late to be useful, but Stable Baselines has an implementation of a replay buffer as part of DQN: https://github.com/hill-a/stable-baselines/blob/master/stable_baselines/deepq/replay_buffer.py
